pkg/jobs/job_info_storage.go

```diff
@@ -0,0 +1,250 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package jobs
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/errors"
+)
+
+// InfoStorage can be used to read and write rows to system.jobs_info table. All
+// operations are scoped under the txn and are are executed on behalf of Job j.
+type InfoStorage struct {
+	j   *Job
+	txn isql.Txn
+}
+
+// InfoStorageWithTxn returns a new InfoStorage with the passed in job and txn.
+func (j *Job) InfoStorageWithTxn(txn isql.Txn) InfoStorage {
+	return InfoStorage{j: j, txn: txn}
+}
+
+// InfoStorageNoTxn returns a new InfoStorage with the passed in job but no txn.
+func (j *Job) InfoStorageNoTxn() InfoStorage {
+	return InfoStorage{j: j}
+}
+
+func (i InfoStorage) checkClaimSession(ctx context.Context) error {
+	row, err := i.txn.QueryRowEx(ctx, "check-claim-session", i.txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		`SELECT claim_session_id FROM system.jobs WHERE id = $1`, i.j.ID())
+	if err != nil {
+		return err
+	}
+
+	if row == nil {
+		return errors.Errorf(
+			"expected session %q for job ID %d but found none", i.j.Session().ID(), i.j.ID())
+	}
+
+	storedSession := []byte(*row[0].(*tree.DBytes))
+	if !bytes.Equal(storedSession, i.j.Session().ID().UnsafeBytes()) {
+		return errors.Errorf(
+			"expected session %q but found %q", i.j.Session().ID(), sqlliveness.SessionID(storedSession))
+	}
+
+	return nil
+}
+
+func (i InfoStorage) get(ctx context.Context, infoKey []byte) ([]byte, bool, error) {
+	if i.txn == nil {
+		var val []byte
+		var exists bool
+		var err error
+		err = i.j.registry.internalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			i.txn = txn
+			val, exists, err = i.get(ctx, infoKey)
+			return err
+		})
+
+		return val, exists, err
+	}
+
+	ctx, sp := tracing.ChildSpan(ctx, "get-job-info")
+	defer sp.Finish()
+
+	j := i.j
+	row, err := i.txn.QueryRowEx(
+		ctx, "job-info-get", i.txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		"SELECT value FROM system.job_info WHERE job_id = $1 AND info_key = $2 ORDER BY written DESC LIMIT 1",
+		j.ID(), infoKey,
+	)
+
+	if err != nil {
+		return nil, false, err
+	}
+
+	if row == nil {
+		return nil, false, nil
+	}
+
+	value, ok := row[0].(*tree.DBytes)
+	if !ok {
+		return nil, false, errors.AssertionFailedf("job info: expected value to be DBytes (was %T)", row[0])
+	}
+
+	return []byte(*value), true, nil
+}
+
+func (i InfoStorage) write(ctx context.Context, infoKey, value []byte) error {
+	if i.txn == nil {
+		return i.j.registry.internalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			i.txn = txn
+			return i.write(ctx, infoKey, value)
```

Comment:
does this leave i.txn non-nil afterwards but with a committed txn?

I'll be honest, I'm not a big fan of this maybe-it-does-maybe-it-doesn't-have-a-txn thing? like, obvious we need a txn -- we make one right here if we don't have one, so why not just make the caller responsible for making/passing a txn all the time?

Reply:
Deleted the NoTxn variant, now all users of InfoStorage must create it with an associated txn that will be used across all operations.